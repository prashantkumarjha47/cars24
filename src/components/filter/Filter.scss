.filters {
  position: relative;
  .form-group {
    margin-bottom: 4rem;
  }
  @media screen and (max-width: 991px) {
    padding-top: 40px;
  }
  .close-btn {
    position: absolute;
    right: 10px;
    top: 10px;
    width: 30px;
    height: 30px;
    padding: 0;
    display: none;
    @media screen and (max-width: 991px) {
      display: block;
    }
  }
  .submit-row {
    text-align: right;
  }
}
