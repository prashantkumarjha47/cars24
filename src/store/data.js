const data = {
  carCopy: [
    {
      id: 1,
      title: "Tata Nexon",
      price: 50000,
      category: "Compact SUV",
      image: "...",
      color: "white",
      condition: "New",
      buildYear: "2020",
      make: "TATA",
      model: "XE",
    },
    /*...*/
    {
      id: 2,
      title: "Kia Sonet",
      price: 225000,
      category: "Compact SUV",
      image: "...",
      color: "black",
      condition: "new",
      buildYear: "2020",
      make: "Kia",
      model: "1.2 HTE 5MT",
    },
    {
      id: 3,
      title: "Ford EcoSport",
      price: 300000,
      category: "Compact SUV",
      image: "...",
      color: "blue",
      condition: "new",
      buildYear: "2020",
      make: "Ford",
      model: "1.5L Petrol Ambiente MT",
    },
    {
      id: 4,
      title: "Hyundai Venue",
      price: 1675000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 5,
      title: "Hyundai Venue",
      price: 875000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 6,
      title: "Hyundai Venue",
      price: 500000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 7,
      title: "Hyundai Venue",
      price: 2575000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 8,
      title: "Hyundai Venue",
      price: 600000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 9,
      title: "Hyundai Venue",
      price: 150000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 10,
      title: "Hyundai Venue",
      price: 480000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 11,
      title: "Hyundai Venue",
      price: 955000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 12,
      title: "Hyundai Venue",
      price: 740000,
      category: "Compact SUV",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 13,
      title: "Hyundai Venue",
      price: 520000,
      category: "Premium hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 15,
      title: "Hyundai Venue",
      price: 98000,
      category: "Premium hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 16,
      title: "Hyundai Venue",
      price: 375000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 17,
      title: "Hyundai Venue",
      price: 347000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 18,
      title: "Hyundai Venue",
      price: 552000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 19,
      title: "Hyundai Venue",
      price: 614000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 20,
      title: "Hyundai Venue",
      price: 1005000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
    {
      id: 21,
      title: "Hyundai Venue",
      price: 75000,
      category: "Hatchback",
      image: "...",
      color: "grey",
      condition: "new",
      buildYear: "2020",
      make: "Hyundai",
      model: "1.2 Kappa MT E",
    },
  ],
  categories: [
    {
      id: 0,
      title: "",
    },
    {
      id: 1,
      title: "Compact SUV",
    },
    {
      id: 2,
      title: "Premium hatchback",
    },
    {
      id: 3,
      title: "Hatchback",
    },
  ],
  condition: [
    {
      id: 0,
      title: "",
    },
    {
      id: 1,
      type: "used",
    },
    {
      id: 2,
      type: "damaged",
    },
  ],
  makes: [
    {
      id: 0,
      title: "",
    },
    {
      id: 1,
      name: "TATA",
    },
    {
      id: 2,
      name: "Kia",
    },
    {
      id: 3,
      name: "Hyundai",
    },
    {
      id: 4,
      name: "Ford",
    },
    {
      id: 5,
      name: "Maruti",
    },
  ],
};

export default data;
