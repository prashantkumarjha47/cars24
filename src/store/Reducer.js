import data from "./data";

const OFFSET = 4;
const arr = data.carCopy.sort((a, b) => a.price - b.price);
const initialState = {
  ...data,
  n: OFFSET,
  cars: data.carCopy.slice(0, OFFSET),
  minVal: arr[0].price,
  maxValu: arr[arr.length - 1].price,
  cart: [],
};

const reducer = (state, action) => {
  switch (action.type) {
    case "GET_CARS":
      const { category, priceRange, vehicleCondition, make, search } = state;
      let filteredCars = state.carCopy
        .filter((car) => {
          return (
            (!category || car.category === category) &&
            (!priceRange || car.price <= priceRange) &&
            (!vehicleCondition || car.vehicleCondition === vehicleCondition) &&
            (!make || car.make === make) &&
            (!search || car.title.toLowerCase().includes(search.toLowerCase()))
          );
        })
        .slice(0, state.n + OFFSET);

      if (state.sort) {
        filteredCars = filteredCars.sort((a, b) =>
          state.sort === "ASC"
            ? parseFloat(a.price) - parseFloat(b.price)
            : parseFloat(b.price) - parseFloat(a.price)
        );
      }

      return {
        ...state,
        n: state.n + OFFSET,
        cars: filteredCars,
      };

    case "GET_CATEGORIES":
      return {
        ...state,
        categories: [...state.categories],
      };
    case "GET_CONDITION":
      return {
        ...state,
        condition: [...state.condition],
      };
    case "SEARCH_CAR":
      return {
        ...state,
        search: action.payload,
        cars: state.cars.filter((car) =>
          car.title?.toLowerCase().includes(action.payload?.toLowerCase())
        ),
      };
    case "SORT_CAR_ASC":
      return {
        ...state,
        sort: "ASC",
        cars: state.cars.sort(
          (a, b) => parseFloat(a.price) - parseFloat(b.price)
        ),
      };
    case "SORT_CAR_DESC":
      return {
        ...state,
        sort: "DESC",
        cars: state.cars.sort(
          (a, b) => parseFloat(b.price) - parseFloat(a.price)
        ),
      };
    case "APPLY_FILTER":
      return {
        ...state,
        n: OFFSET,
        ...action.payload,
        cars: [],
      };
    case "BOOK_NOW":
      return {
        ...state,
        cart: [...state.cart, action.payload],
      };

    default:
      return state;
  }
};

export { reducer, initialState };
