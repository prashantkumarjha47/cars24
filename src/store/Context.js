import React, { useReducer } from "react";
import { reducer, initialState } from "./Reducer";

const Context = React.createContext();

const ContextProvider = (props) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <Context.Provider value={[state, dispatch]}>
      {props.children}
    </Context.Provider>
  );
};

export { ContextProvider, Context };
