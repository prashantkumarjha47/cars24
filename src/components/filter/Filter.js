import React, { useContext, useState } from "react";
import { Input, Form, FormGroup, Label, Container, Button } from "reactstrap";
import InputRange from "react-input-range";
import { Context } from "../../store/Context";
import "react-input-range/lib/css/index.css";
import "./Filter.scss";
import closeIcon from "../../assets/images/close-icon.svg";

function Filter(props) {
  const [state, dispatch] = useContext(Context);
  const [priceRange, setPriceRange] = useState(state.minVal);
  const [category, setCategories] = useState("");
  const [vehicleCondition, setVehicleCondition] = useState("");
  const [make, setmake] = useState("");

  const onApplyFilter = () => {
    const filters = {
      priceRange,
      category,
      vehicleCondition,
      make,
    };

    updateFilterInStore(filters);
  };

  const onResetFilter = () => {
    setPriceRange(0);
    setmake("");
    setVehicleCondition("");
    setCategories("");

    const filters = {
      priceRange: 0,
      category: "",
      vehicleCondition: "",
      make: "",
    };

    updateFilterInStore(filters);
  };

  const updateFilterInStore = (filters) => {
    dispatch({ type: "APPLY_FILTER", payload: filters });
    dispatch({ type: "GET_CARS" });
  };
  console.log(state);
  return (
    <div className="filters">
      <Button color="link" className="close-btn" onClick={props.onCloseClick}>
        <img src={closeIcon} alt="Close" />
      </Button>
      <Container>
        <Form>
          <FormGroup>
            <Label for="exampleSelect">Category</Label>
            <Input
              type="select"
              name="select"
              id="exampleSelect"
              value={category}
              onChange={(e) => setCategories(e.target.value)}
            >
              {state.categories.map((cat) => (
                <option key={cat.id} value={cat.title}>
                  {cat.title}
                </option>
              ))}
            </Input>
          </FormGroup>
          <FormGroup>
            <Label for="exampleSelect">Price range</Label>
            <InputRange
              maxValue={state.maxValu}
              minValue={state.minVal}
              value={priceRange}
              onChange={(value) => setPriceRange(value)}
            />
          </FormGroup>
          <FormGroup>
            <Label for="exampleSelect">Vehicle Condition</Label>
            <Input
              onChange={(e) => setVehicleCondition(e.target.value)}
              value={vehicleCondition}
              type="select"
              name="select"
              id="exampleSelect"
            >
              {state.condition.map((cond) => (
                <option key={cond.id} value={cond.type}>
                  {cond.type}
                </option>
              ))}
            </Input>
          </FormGroup>
          <FormGroup>
            <Label for="exampleSelect">Car make</Label>
            <Input
              onChange={(e) => setmake(e.target.value)}
              type="select"
              name="select"
              value={make}
              id="exampleSelect"
            >
              {state.makes.map((make) => (
                <option key={make.id} value={make.name}>
                  {make.name}
                </option>
              ))}
            </Input>
          </FormGroup>
          <FormGroup className="submit-row">
            <Button onClick={onApplyFilter}>Apply</Button>{" "}
            <Button onClick={onResetFilter}>Reset</Button>
          </FormGroup>
        </Form>
      </Container>
    </div>
  );
}

export default React.memo(Filter);
