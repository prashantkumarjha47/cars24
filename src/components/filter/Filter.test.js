import renderer from "react-test-renderer";
import Filter from "./Filter";
import { ContextProvider } from "../../store/Context";

test("Filter component loaded", () => {
  const component = renderer.create(
    <ContextProvider>
      <Filter onCloseClick={() => {}} />
    </ContextProvider>
  );
  expect(component.toJSON()).toMatchSnapshot();
});
