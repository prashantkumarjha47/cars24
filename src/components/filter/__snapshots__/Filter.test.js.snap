// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component loaded 1`] = `
<div
  className="filters"
>
  <button
    aria-label={null}
    className="close-btn btn btn-link"
    onClick={[Function]}
    type="button"
  >
    <img
      alt="Close"
      src="close-icon.svg"
    />
  </button>
  <div
    className="container"
  >
    <form
      className=""
    >
      <div
        className="form-group"
      >
        <label
          className=""
          htmlFor="exampleSelect"
        >
          Category
        </label>
        <select
          className="form-control"
          id="exampleSelect"
          name="select"
          onChange={[Function]}
          value=""
        >
          <option
            value=""
          >
            
          </option>
          <option
            value="Compact SUV"
          >
            Compact SUV
          </option>
          <option
            value="Premium hatchback"
          >
            Premium hatchback
          </option>
          <option
            value="Hatchback"
          >
            Hatchback
          </option>
        </select>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
          htmlFor="exampleSelect"
        >
          Price range
        </label>
        <div
          aria-disabled={false}
          className="input-range"
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
        >
          <span
            className="input-range__label input-range__label--min"
          >
            <span
              className="input-range__label-container"
            >
              0
            </span>
          </span>
          <div
            className="input-range__track input-range__track--background"
            onMouseDown={[Function]}
            onTouchStart={[Function]}
          >
            <div
              className="input-range__track input-range__track--active"
              style={
                Object {
                  "left": "0%",
                  "width": "0%",
                }
              }
            />
            <span
              className="input-range__slider-container"
              style={
                Object {
                  "left": "0%",
                  "position": "absolute",
                }
              }
            >
              <span
                className="input-range__label input-range__label--value"
              >
                <span
                  className="input-range__label-container"
                >
                  0
                </span>
              </span>
              <div
                aria-valuemax={20}
                aria-valuemin={0}
                aria-valuenow={0}
                className="input-range__slider"
                draggable="false"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onTouchStart={[Function]}
                role="slider"
                tabIndex="0"
              />
            </span>
          </div>
          <span
            className="input-range__label input-range__label--max"
          >
            <span
              className="input-range__label-container"
            >
              20
            </span>
          </span>
        </div>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
          htmlFor="exampleSelect"
        >
          Vehicle Condition
        </label>
        <select
          className="form-control"
          id="exampleSelect"
          name="select"
          onChange={[Function]}
          value=""
        >
          <option
            value="used"
          >
            used
          </option>
          <option
            value="damaged"
          >
            damaged
          </option>
        </select>
      </div>
      <div
        className="form-group"
      >
        <label
          className=""
          htmlFor="exampleSelect"
        >
          Car make
        </label>
        <select
          className="form-control"
          id="exampleSelect"
          name="select"
          onChange={[Function]}
          value=""
        >
          <option
            value="TATA"
          >
            TATA
          </option>
          <option
            value="Kia"
          >
            Kia
          </option>
          <option
            value="Hyundai"
          >
            Hyundai
          </option>
          <option
            value="Ford"
          >
            Ford
          </option>
          <option
            value="Maruti"
          >
            Maruti
          </option>
        </select>
      </div>
      <div
        className="submit-row form-group"
      >
        <button
          aria-label={null}
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Apply
        </button>
         
        <button
          aria-label={null}
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
</div>
`;
